sql/ir_generator.go

```diff
@@ -0,0 +1,375 @@
+				fc, err := parseFeatureColumn(&colExpr.sexp)
+				if err != nil {
+					return nil, fmt.Errorf("parse FeatureColumn failed: %v", err)
+				}
+				fcList = append(fcList, fc)
+			}
+		}
+		fcMap[target] = fcList
+	}
+	label := &codegen.LabelColumn{
+		FieldMeta: &codegen.FieldMeta{
+			Name: tc.label,
+		}}
+
+	// TODO(typhoonzero): fill in ValidationSelect when VALIDATE clause is ready
+	return &codegen.TrainIR{
+		DataSource:       connStr,
+		Select:           slct.standardSelect.String(),
+		ValidationSelect: "",
+		Estimator:        estimator,
+		Attributes:       attrList,
+		Features:         fcMap,
+		Label:            label,
+	}, nil
+}
+
+func generateAttributeIR(attrs *attrs) ([]*codegen.Attribute, error) {
+	ret := []*codegen.Attribute{}
+	for k, v := range *attrs {
+		resolvedExpr, err := parseExpression(v)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, &codegen.Attribute{
+			Key:   k,
+			Value: resolvedExpr,
+		})
+	}
+	return ret, nil
+}
+
+// -------------------------- expression parsers --------------------------------------
+
+// parseExpression is a recursive function, it returns the actual value
+// or struct the expression stands for, e.g.
+// 1                 ->  int(1)
+// "string"          ->  string("string")
+// [1,2,3]           ->  []int{1,2,3}
+// NUMERIC(c1)       ->  &codegen.NumericColumn{Key: "c1"...}
+// [NUMERIC(c1), c2] ->  [&codegen.NumericColumn{Key: "c1"...}, string("c2")]
+//
+// parameter e could be type `*expr` or `*exprlist` for recursive call.
+func parseExpression(e interface{}) (interface{}, error) {
+	if expr, ok := e.(*expr); ok {
+		if expr.typ != 0 {
+			// TODO(typhoonzero): infer the element expression type like int, float, string
+			return expr.val, nil
+		}
+		return parseExpression(&expr.sexp)
+	}
+	el, ok := e.(*exprlist)
+	if !ok {
+		return nil, fmt.Errorf("input of parseExpression must be `expr` or `exprlist` given %s", e)
+	}
+
+	headTyp := (*el)[0].typ
+	if headTyp == IDENT {
+		// expression is a function call format like `NUMERIC(c1)`
+		return parseFeatureColumn(el)
+	} else if headTyp == '[' {
+		// expression is a list of things
+		var list []interface{}
+		for idx, expr := range *el {
+			if idx > 0 {
+				if expr.sexp == nil {
+					intVal, err := strconv.Atoi(expr.val)
+					// TODO(typhoonzero): support list of float etc.
+					if err != nil {
+						list = append(list, expr.val)
+					} else {
+						list = append(list, intVal)
+					}
+				} else {
+					value, err := parseExpression(&expr.sexp)
+					if err != nil {
+						return nil, err
+					}
+					list = append(list, value)
+				}
+			}
+		}
+		return list, nil
+	}
+	return nil, fmt.Errorf("not supported expr: %v", el)
+}
+
+func parseFeatureColumn(el *exprlist) (codegen.FeatureColumn, error) {
+	head := (*el)[0].val
+	if head == "" {
+		return nil, fmt.Errorf("column description expects format like NUMERIC(key) etc, got %v", el)
+	}
+
+	switch strings.ToUpper(head) {
+	// case dense:
```

Comment:
Are we supporting [`NUMERIC(field, n[, delimiter=comma])`](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/user_guide.md#numeric) as the equivalent for `DENSE`?

Reply:
No, we'll only support below methods:

`NUMERIC(DENSE(field, shape, delimiter))` or

`NUMERIC(field)` and infer the field's shape and delimiter using feature derivation, or

`field` or nothing for full feature derivation.